### ISSUE TITLE
Unnecessary synchronization of deterministic variable in Mirrored Distributed mode

### ISSUE THREAD

**Issue by olegmyrk** [tensorflow/tensorflow#35347]
**System information**
* Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
* OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04.3 LTS
* TensorFlow installed from (source or binary): binary
* TensorFlow version (use command below): v2.1.0-rc0-47-g064e153 2.1.0-rc1 (python3 -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)")
* Python version: Python 3.6.8
* CUDA/cuDNN version: Driver Version: 440.33.01, CUDA Version: 10.2, cuDNN 7.6.2
GPU model and memory: Tesla V100-SXM2-16GB

**Describe the current behavior**
I have a custom implementation of spectral normalization:
https://github.com/olegmyrk/SPADE-Tensorflow/blob/4203f30b6253a9d4743962087896fab26381c67b/ops.py#L579

It defines a non-trainable variable
```
u = tf.compat.v1.get_variable("u", [1, w_shape[-1]], initializer=tf.compat.v1.random_normal_initializer(), trainable=False, aggregation=tf.compat.v2.VariableAggregation.ONLY_FIRST_REPLICA)
```
that is used for power iteration
https://en.wikipedia.org/wiki/Power_iteration
It is updated using deterministic computation (ie it depends only on weight matrix, but not batch inputs) at each gradient step:
https://github.com/olegmyrk/SPADE-Tensorflow/blob/4203f30b6253a9d4743962087896fab26381c67b/ops.py#L606

Currently in mirrored distributed mode this adds a very expensive synchronization step for this variable.

I could set 
```
aggregation=tf.compat.v2.VariableAggregation.NONE
```
but that doesn't work in distributed mode, according to manual:
https://www.tensorflow.org/api_docs/python/tf/VariableAggregation
> NONE: This is the default, giving an error if you use a variable-update operation with multiple replicas.

Additionally it takes ~40 minutes to build the graph with spectral normalization in distributed mirrored mode (with multiple GPUs) as opposed to 20 minutes without spectral normalization enabled.

Although this might be a generic issue related to slow AutoGraph in Mirrored Distributed mode:
https://github.com/tensorflow/tensorflow/issues/35346

**Describe the expected behavior**
I need to specify that there is no need to synchronize this variable as it has the same deterministic value on all replicas.

**Code to reproduce the issue**
https://github.com/olegmyrk/SPADE-Tensorflow/blob/4203f30b6253a9d4743962087896fab26381c67b/ops.py#L606

**Other info / logs**
Using `NCCL_DEBUG=INFO NCCL_DEBUG_SUBSYS=ALL`:

*Without spectral norm*
INFO:tensorflow:batch_all_reduce: 219 all-reduces with algorithm = nccl, num_packs = 1, agg_small_grads_max_bytes = 0 and agg_small_grads_max_group = 10
INFO:tensorflow:batch_all_reduce: 1058 all-reduces with algorithm = nccl, num_packs = 1, agg_small_grads_max_bytes = 0 and agg_small_grads_max_group = 10
INFO:tensorflow:batch_all_reduce: 2150 all-reduces with algorithm = nccl, num_packs = 1, agg_small_grads_max_bytes = 0 and agg_small_grads_max_group = 10
INFO:tensorflow:batch_all_reduce: 4402 all-reduces with algorithm = nccl, num_packs = 1, agg_small_grads_max_bytes = 0 and agg_small_grads_max_group = 10

INFO:tensorflow:batch_all_reduce: 219 all-reduces with algorithm = nccl, num_packs = 1, agg_small_grads_max_bytes = 0 and agg_small_grads_max_group = 10
INFO:tensorflow:batch_all_reduce: 1058 all-reduces with algorithm = nccl, num_packs = 1, agg_small_grads_max_bytes = 0 and agg_small_grads_max_group = 10
INFO:tensorflow:batch_all_reduce: 2150 all-reduces with algorithm = nccl, num_packs = 1, agg_small_grads_max_bytes = 0 and agg_small_grads_max_group = 10
INFO:tensorflow:batch_all_reduce: 4402 all-reduces with algorithm = nccl, num_packs = 1, agg_small_grads_max_bytes = 0 and agg_small_grads_max_group = 10

*With spectral norm*
INFO:tensorflow:batch_all_reduce: 459 all-reduces with algorithm = nccl, num_packs = 1, agg_small_grads_max_bytes = 0 and agg_small_grads_max_group = 10
INFO:tensorflow:batch_all_reduce: 1666 all-reduces with algorithm = nccl, num_packs = 1, agg_small_grads_max_bytes = 0 and agg_small_grads_max_group = 10
INFO:tensorflow:batch_all_reduce: 2662 all-reduces with algorithm = nccl, num_packs = 1, agg_small_grads_max_bytes = 0 and agg_small_grads_max_group = 10
INFO:tensorflow:batch_all_reduce: 6610 all-reduces with algorithm = nccl, num_packs = 1, agg_small_grads_max_bytes = 0 and agg_small_grads_max_group = 10

INFO:tensorflow:batch_all_reduce: 459 all-reduces with algorithm = nccl, num_packs = 1, agg_small_grads_max_bytes = 0 and agg_small_grads_max_group = 10
INFO:tensorflow:batch_all_reduce: 1666 all-reduces with algorithm = nccl, num_packs = 1, agg_small_grads_max_bytes = 0 and agg_small_grads_max_group = 10
INFO:tensorflow:batch_all_reduce: 2662 all-reduces with algorithm = nccl, num_packs = 1, agg_small_grads_max_bytes = 0 and agg_small_grads_max_group = 10
INFO:tensorflow:batch_all_reduce: 6610 all-reduces with algorithm = nccl, num_packs = 1, agg_small_grads_max_bytes = 0 and agg_small_grads_max_group = 10

**Comment by anj-s** [comment]
olegmyrk@ Can you create a variable with synchronization=ON_READ and aggregation=ONLY_FIRST_REPLICA? This will allow you to create variables that are not synchronized every time you update them. They are called SyncOnRead variables and are synchronized when you read the variable. By specifying ONLY_FIRST_REPLICA you can ensure that you only read from the first device. An example of this can be found in TF [metrics](https://github.com/tensorflow/tensorflow/blob/r2.0/tensorflow/python/keras/metrics.py#L245). It uses a different aggregation though.

**Comment by olegmyrk** [comment]
Do you think it will help if I read and update the variables on every training step? It's just that the variables are deterministic functions of network weights, but not inputs.

**Comment by anj-s** [comment]
olegmyrk@ Using MirroredVariables with ONLY_FIRST_REPLICA should have resulted in a broadcast of values from the first replica to all other replicas when you called `u.assign(u_hat)`. Using SyncOnRead variables will help you avoid this broadcast of values. However I think I am not sure how this is resulting in additional allreduce calls. Are we increasing the number of trainable variables somehow or taking additional gradients? can we confirm that the set of trainable variables remain the same?

**Comment by olegmyrk** [comment]
Thanks, this resolved this issue for me.

**Comment by anj-s** [comment]
@olegmyrk How did you resolve the additional allreduce calls? Did you have additional trainable variables?

**Comment by olegmyrk** [comment]
I just applied your suggested change and empirically I don't see performance issues with spectral norm anymore. To be fair I also did a bunch of other changes in between. Anyway, thanks!

**Comment by anj-s** [comment]
@olegmyrk Thanks for responding and letting me know. Glad the issue is fixed!
